{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9930aa00",
   "metadata": {},
   "source": [
    "# Demo of Utilizing New GraphQL API for SNP Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "971a2377-b5c8-467b-8ec6-245b0bc2367a",
   "metadata": {},
   "source": [
    "## Setting Up API Connection\n",
    "\n",
    "The script begins by setting up a connection to a GraphQL API, specifying the base URL and the endpoint."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "99cc287f-558c-4cba-9610-d6cd82a480f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import pandas as pd\n",
    "from config.settings import settings\n",
    "\n",
    "BASE_URL = settings.API_URL\n",
    "GRAPHQL_ENDPOINT = 'graphql'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61817c4d-50ad-48e8-86c6-9093bc1f6048",
   "metadata": {},
   "source": [
    "## Understanding Annotations in the API\n",
    "\n",
    "It executes a GET request to retrieve a list of annotations. These annotations describe various data fields available through the API, detailing their characteristics such as name, description, and how they relate to each other in a hierarchical manner, much like a structured catalog of options you can query. \n",
    "\n",
    "**api_field:** Specifies the field name as it should be used in API requests, particularly when crafting queries for a GraphQL API. This ensures you're asking for data in a format the API understands."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "88e24db8-b2c3-40f6-839e-895f57c88609",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>leaf</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "      <th>sort</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>detail</th>\n",
       "      <th>link</th>\n",
       "      <th>pmid</th>\n",
       "      <th>field_type</th>\n",
       "      <th>keyword_searchable</th>\n",
       "      <th>api_field</th>\n",
       "      <th>root_url</th>\n",
       "      <th>sample_url</th>\n",
       "      <th>value_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>root</td>\n",
       "      <td>Annotation</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Basic Info</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Basic information about the variant, such as c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>26</td>\n",
       "      <td>False</td>\n",
       "      <td>ANNOVAR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Pre-computed ANNOVAR annotations for all alter...</td>\n",
       "      <td>http://annovar.openbioinformatics.org/en/lates...</td>\n",
       "      <td>20601685</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>208</td>\n",
       "      <td>False</td>\n",
       "      <td>SnpEff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>AnpEff is a program for annotating and predict...</td>\n",
       "      <td>http://pcingola.github.io/SnpEff/</td>\n",
       "      <td>22728672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>132</td>\n",
       "      <td>False</td>\n",
       "      <td>VEP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Variant Effect Predictor (VEP) is developed by...</td>\n",
       "      <td>https://uswest.ensembl.org/info/docs/tools/vep...</td>\n",
       "      <td>27268795</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>621</td>\n",
       "      <td>True</td>\n",
       "      <td>sno_miRNA_type</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>495</td>\n",
       "      <td>the type of snoRNA or miRNA (from miRBase/snoR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>text</td>\n",
       "      <td>True</td>\n",
       "      <td>sno_miRNA_type</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>622</td>\n",
       "      <td>True</td>\n",
       "      <td>splicing_consensus_ada_score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>495</td>\n",
       "      <td>splicing-change prediction for splicing consen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>float</td>\n",
       "      <td>NaN</td>\n",
       "      <td>splicing_consensus_ada_score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>623</td>\n",
       "      <td>True</td>\n",
       "      <td>splicing_consensus_rf_score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>495</td>\n",
       "      <td>splicing-change prediction for splicing consen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>float</td>\n",
       "      <td>NaN</td>\n",
       "      <td>splicing_consensus_rf_score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>624</td>\n",
       "      <td>True</td>\n",
       "      <td>target_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>495</td>\n",
       "      <td>target gene (for promoter, enhancer, etc.) bas...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>text</td>\n",
       "      <td>True</td>\n",
       "      <td>target_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>625</td>\n",
       "      <td>True</td>\n",
       "      <td>ultra_sensitive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>495</td>\n",
       "      <td>whether defined as ultra-sensitive region base...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>text</td>\n",
       "      <td>True</td>\n",
       "      <td>ultra_sensitive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>546 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      id   leaf                          name       label  sort parent_id  \\\n",
       "0      0  False                          root  Annotation   0.0       NaN   \n",
       "1      1  False                    Basic Info         NaN   1.0         0   \n",
       "2     26  False                       ANNOVAR         NaN   2.0         0   \n",
       "3    208  False                        SnpEff         NaN   3.0         0   \n",
       "4    132  False                           VEP         NaN   4.0         0   \n",
       "..   ...    ...                           ...         ...   ...       ...   \n",
       "541  621   True                sno_miRNA_type         NaN   NaN       495   \n",
       "542  622   True  splicing_consensus_ada_score         NaN   NaN       495   \n",
       "543  623   True   splicing_consensus_rf_score         NaN   NaN       495   \n",
       "544  624   True                   target_gene         NaN   NaN       495   \n",
       "545  625   True               ultra_sensitive         NaN   NaN       495   \n",
       "\n",
       "                                                detail  \\\n",
       "0                                                  NaN   \n",
       "1    Basic information about the variant, such as c...   \n",
       "2    Pre-computed ANNOVAR annotations for all alter...   \n",
       "3    AnpEff is a program for annotating and predict...   \n",
       "4    Variant Effect Predictor (VEP) is developed by...   \n",
       "..                                                 ...   \n",
       "541  the type of snoRNA or miRNA (from miRBase/snoR...   \n",
       "542  splicing-change prediction for splicing consen...   \n",
       "543  splicing-change prediction for splicing consen...   \n",
       "544  target gene (for promoter, enhancer, etc.) bas...   \n",
       "545  whether defined as ultra-sensitive region base...   \n",
       "\n",
       "                                                  link      pmid field_type  \\\n",
       "0                                                  NaN       NaN        NaN   \n",
       "1                                                  NaN       NaN        NaN   \n",
       "2    http://annovar.openbioinformatics.org/en/lates...  20601685        NaN   \n",
       "3                    http://pcingola.github.io/SnpEff/  22728672        NaN   \n",
       "4    https://uswest.ensembl.org/info/docs/tools/vep...  27268795        NaN   \n",
       "..                                                 ...       ...        ...   \n",
       "541                                                NaN       NaN       text   \n",
       "542                                                NaN       NaN      float   \n",
       "543                                                NaN       NaN      float   \n",
       "544                                                NaN       NaN       text   \n",
       "545                                                NaN       NaN       text   \n",
       "\n",
       "    keyword_searchable                     api_field root_url sample_url  \\\n",
       "0                  NaN                           NaN      NaN        NaN   \n",
       "1                  NaN                           NaN      NaN        NaN   \n",
       "2                  NaN                           NaN      NaN        NaN   \n",
       "3                  NaN                           NaN      NaN        NaN   \n",
       "4                  NaN                           NaN      NaN        NaN   \n",
       "..                 ...                           ...      ...        ...   \n",
       "541               True                sno_miRNA_type      NaN        NaN   \n",
       "542                NaN  splicing_consensus_ada_score      NaN        NaN   \n",
       "543                NaN   splicing_consensus_rf_score      NaN        NaN   \n",
       "544               True                   target_gene      NaN        NaN   \n",
       "545               True               ultra_sensitive      NaN        NaN   \n",
       "\n",
       "    value_type  \n",
       "0          NaN  \n",
       "1          NaN  \n",
       "2          NaN  \n",
       "3          NaN  \n",
       "4          NaN  \n",
       "..         ...  \n",
       "541        NaN  \n",
       "542        NaN  \n",
       "543        NaN  \n",
       "544        NaN  \n",
       "545        NaN  \n",
       "\n",
       "[546 rows x 15 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response = requests.get(f\"{BASE_URL}annotations\")\n",
    "\n",
    "annotations = response.json()\n",
    "annotations_df = pd.DataFrame(annotations['results'])\n",
    "annotations_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f136d8e2-6831-4a5e-a046-c52dcf5aea24",
   "metadata": {},
   "source": [
    "## Extracting SNP Data Through a GraphQL Query\n",
    "\n",
    "The script continues by constructing a GraphQL query designed to fetch specific information about Single Nucleotide Polymorphisms (SNPs) based on criteria such as chromosome number and position range. This query illustrates GraphQL's capability to precisely target and retrieve the needed data from the server, thus optimizing the data acquisition process. The response from this query provides detailed attributes of SNPs for subsequent processing or analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d755f465-1e12-48b3-83fd-b86266944984",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"\"\"\n",
    "query MyQuery {\n",
    "  GetSNPsByChromosome(chr: \"1\", end: 1000000, start: 10) {\n",
    "    alt \n",
    "    chr \n",
    "    pos\n",
    "    rs_dbSNP151\n",
    "    ref\n",
    "    ANNOVAR_ensembl_Effect \n",
    "    ANNOVAR_refseq_Effect\n",
    "  }\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "response = requests.post(f\"{BASE_URL}{GRAPHQL_ENDPOINT}\", json={'query': query})\n",
    "\n",
    "data = json.loads(response.text)\n",
    "snps_by_chromosome = data['data']['GetSNPsByChromosome']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eec7709-6fbe-4d8d-a98f-1e1c6f64af05",
   "metadata": {},
   "source": [
    "## Processing and Displaying the Data\n",
    "\n",
    "After receiving data from the GraphQL query, the script processes it for analysis. This involves flattening the nested structure of the data response to a more straightforward, table-like format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "81967952-8531-48e4-8767-b7a2b9c22b04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alt</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>rs_dbSNP151</th>\n",
       "      <th>ref</th>\n",
       "      <th>ANNOVAR_ensembl_Effect</th>\n",
       "      <th>ANNOVAR_refseq_Effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>54353</td>\n",
       "      <td>rs140052487</td>\n",
       "      <td>C</td>\n",
       "      <td>ncRNA_intronic|downstream</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>54564</td>\n",
       "      <td>rs558796213</td>\n",
       "      <td>G</td>\n",
       "      <td>ncRNA_intronic|downstream</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>54591</td>\n",
       "      <td>rs561234294</td>\n",
       "      <td>A</td>\n",
       "      <td>ncRNA_intronic|downstream</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>16071</td>\n",
       "      <td>rs541172944</td>\n",
       "      <td>G</td>\n",
       "      <td>ncRNA_intronic|downstream</td>\n",
       "      <td>ncRNA_intronic|downstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>16141</td>\n",
       "      <td>rs529651976</td>\n",
       "      <td>C</td>\n",
       "      <td>ncRNA_intronic|downstream</td>\n",
       "      <td>ncRNA_intronic|downstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>16280</td>\n",
       "      <td>rs866639523</td>\n",
       "      <td>T</td>\n",
       "      <td>ncRNA_intronic|downstream</td>\n",
       "      <td>ncRNA_intronic|downstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>13380</td>\n",
       "      <td>rs571093408</td>\n",
       "      <td>C</td>\n",
       "      <td>ncRNA_exonic|ncRNA_intronic|downstream</td>\n",
       "      <td>ncRNA_exonic|downstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>49298</td>\n",
       "      <td>rs10399793</td>\n",
       "      <td>T</td>\n",
       "      <td>upstream</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>54676</td>\n",
       "      <td>rs2462492</td>\n",
       "      <td>C</td>\n",
       "      <td>ncRNA_intronic|downstream</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  alt chr    pos  rs_dbSNP151 ref                  ANNOVAR_ensembl_Effect  \\\n",
       "0   A   1  54353  rs140052487   C               ncRNA_intronic|downstream   \n",
       "1   T   1  54564  rs558796213   G               ncRNA_intronic|downstream   \n",
       "2   G   1  54591  rs561234294   A               ncRNA_intronic|downstream   \n",
       "3   A   1  16071  rs541172944   G               ncRNA_intronic|downstream   \n",
       "4   T   1  16141  rs529651976   C               ncRNA_intronic|downstream   \n",
       "5   C   1  16280  rs866639523   T               ncRNA_intronic|downstream   \n",
       "6   G   1  13380  rs571093408   C  ncRNA_exonic|ncRNA_intronic|downstream   \n",
       "7   C   1  49298   rs10399793   T                                upstream   \n",
       "8   T   1  54676    rs2462492   C               ncRNA_intronic|downstream   \n",
       "\n",
       "       ANNOVAR_refseq_Effect  \n",
       "0                 intergenic  \n",
       "1                 intergenic  \n",
       "2                 intergenic  \n",
       "3  ncRNA_intronic|downstream  \n",
       "4  ncRNA_intronic|downstream  \n",
       "5  ncRNA_intronic|downstream  \n",
       "6    ncRNA_exonic|downstream  \n",
       "7                 intergenic  \n",
       "8                 intergenic  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flattened_data = pd.json_normalize(snps_by_chromosome)\n",
    "snp_df = pd.DataFrame(flattened_data)\n",
    "snp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9e69d00",
   "metadata": {},
   "source": [
    "# Get SNP by VCF File\n",
    "This is done using the IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c66f99af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': '2:10662G>C',\n",
       "  'chr': '2',\n",
       "  'pos': 10662,\n",
       "  'ANNOVAR_ensembl_Effect': 'intergenic'},\n",
       " {'id': '2:10632C>A',\n",
       "  'chr': '2',\n",
       "  'pos': 10632,\n",
       "  'ANNOVAR_ensembl_Effect': 'intergenic'}]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "query myQuery {\n",
    "  GetSNPsByIDs(\n",
    "    ids: [\"2:10662G>C\", \"2:10632C>A\"]\n",
    "    page_args: {from_: 0, size: 5}\n",
    "  ) {\n",
    "    id\n",
    "    chr\n",
    "    pos\n",
    "    ANNOVAR_ensembl_Effect\n",
    "  }\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "response = requests.post(f\"{BASE_URL}{GRAPHQL_ENDPOINT}\", json={'query': query})\n",
    "\n",
    "data = json.loads(response.text)\n",
    "snps_by_id = data['data']['GetSNPsByIDs']\n",
    "snps_by_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c01ae00e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ANNOVAR_ensembl_Effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2:10662G&gt;C</td>\n",
       "      <td>2</td>\n",
       "      <td>10662</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2:10632C&gt;A</td>\n",
       "      <td>2</td>\n",
       "      <td>10632</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           id chr    pos ANNOVAR_ensembl_Effect\n",
       "0  2:10662G>C   2  10662             intergenic\n",
       "1  2:10632C>A   2  10632             intergenic"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flattened_data = pd.json_normalize(snps_by_id)\n",
    "snp_df = pd.DataFrame(flattened_data)\n",
    "snp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48afc5bc",
   "metadata": {},
   "source": [
    "# Get SNP by RsID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e0c83fa6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'rs_dbSNP151': 'rs189126619', 'chr': '2'}]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "query myQuery {\n",
    "  GetSNPsByRsID(rsID: \"rs189126619\", filter_args: {exists: [\"rs_dbSNP151\"]}) {\n",
    "    rs_dbSNP151\n",
    "    chr\n",
    "  }\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "response = requests.post(f\"{BASE_URL}{GRAPHQL_ENDPOINT}\", json={'query': query})\n",
    "\n",
    "data = json.loads(response.text)\n",
    "snps_by_RsID = data['data']['GetSNPsByRsID']\n",
    "snps_by_RsID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "80dd848c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs_dbSNP151</th>\n",
       "      <th>chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs189126619</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   rs_dbSNP151 chr\n",
       "0  rs189126619   2"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flattened_data = pd.json_normalize(snps_by_RsID)\n",
    "snp_df = pd.DataFrame(flattened_data)\n",
    "snp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bbd1b60",
   "metadata": {},
   "source": [
    "# Get SNP by RsIDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4ea36768",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'rs_dbSNP151': 'rs115366554', 'chr': '2'},\n",
       " {'rs_dbSNP151': 'rs189126619', 'chr': '2'}]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "query myQuery {\n",
    "  GetSNPsByRsIDs(rsIDs: [\"rs115366554\", \"rs189126619\"], filter_args: {exists: [\"rs_dbSNP151\"]}) {\n",
    "    rs_dbSNP151\n",
    "    chr\n",
    "  }\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "response = requests.post(f\"{BASE_URL}{GRAPHQL_ENDPOINT}\", json={'query': query})\n",
    "\n",
    "data = json.loads(response.text)\n",
    "snps_by_RsIDs = data['data']['GetSNPsByRsIDs']\n",
    "snps_by_RsIDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "5f1d6aaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs_dbSNP151</th>\n",
       "      <th>chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs115366554</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs189126619</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   rs_dbSNP151 chr\n",
       "0  rs115366554   2\n",
       "1  rs189126619   2"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flattened_data = pd.json_normalize(snps_by_RsIDs)\n",
    "snp_df = pd.DataFrame(flattened_data)\n",
    "snp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3426cb80",
   "metadata": {},
   "source": [
    "# Get aggregates by chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d7814ed9-7f03-44b9-a93b-8b0102b2dbc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr': {'histogram': [{'key': '9786.54', 'doc_count': 9},\n",
       "   {'key': '14679.810000000001', 'doc_count': 0},\n",
       "   {'key': '19573.08', 'doc_count': 0},\n",
       "   {'key': '24466.350000000002', 'doc_count': 0},\n",
       "   {'key': '29359.620000000003', 'doc_count': 0},\n",
       "   {'key': '34252.89', 'doc_count': 0},\n",
       "   {'key': '39146.16', 'doc_count': 0},\n",
       "   {'key': '44039.43000000001', 'doc_count': 0},\n",
       "   {'key': '48932.700000000004', 'doc_count': 0},\n",
       "   {'key': '53825.97', 'doc_count': 0},\n",
       "   {'key': '58719.240000000005', 'doc_count': 0},\n",
       "   {'key': '63612.51000000001', 'doc_count': 0},\n",
       "   {'key': '68505.78', 'doc_count': 0},\n",
       "   {'key': '73399.05', 'doc_count': 0},\n",
       "   {'key': '78292.32', 'doc_count': 0},\n",
       "   {'key': '83185.59000000001', 'doc_count': 0},\n",
       "   {'key': '88078.86000000002', 'doc_count': 0},\n",
       "   {'key': '92972.13', 'doc_count': 0},\n",
       "   {'key': '97865.40000000001', 'doc_count': 0},\n",
       "   {'key': '102758.67000000001', 'doc_count': 0},\n",
       "   {'key': '107651.94', 'doc_count': 0},\n",
       "   {'key': '112545.21', 'doc_count': 0},\n",
       "   {'key': '117438.48000000001', 'doc_count': 0},\n",
       "   {'key': '122331.75000000001', 'doc_count': 0},\n",
       "   {'key': '127225.02000000002', 'doc_count': 0},\n",
       "   {'key': '132118.29', 'doc_count': 0},\n",
       "   {'key': '137011.56', 'doc_count': 0},\n",
       "   {'key': '141904.83000000002', 'doc_count': 0},\n",
       "   {'key': '146798.1', 'doc_count': 0},\n",
       "   {'key': '151691.37000000002', 'doc_count': 0},\n",
       "   {'key': '156584.64', 'doc_count': 0},\n",
       "   {'key': '161477.91', 'doc_count': 0},\n",
       "   {'key': '166371.18000000002', 'doc_count': 0},\n",
       "   {'key': '171264.45', 'doc_count': 0},\n",
       "   {'key': '176157.72000000003', 'doc_count': 0},\n",
       "   {'key': '181050.99000000002', 'doc_count': 0},\n",
       "   {'key': '185944.26', 'doc_count': 0},\n",
       "   {'key': '190837.53000000003', 'doc_count': 0},\n",
       "   {'key': '195730.80000000002', 'doc_count': 0},\n",
       "   {'key': '200624.07', 'doc_count': 0},\n",
       "   {'key': '205517.34000000003', 'doc_count': 0},\n",
       "   {'key': '210410.61000000002', 'doc_count': 0},\n",
       "   {'key': '215303.88', 'doc_count': 0},\n",
       "   {'key': '220197.15000000002', 'doc_count': 0},\n",
       "   {'key': '225090.42', 'doc_count': 0},\n",
       "   {'key': '229983.69000000003', 'doc_count': 0},\n",
       "   {'key': '234876.96000000002', 'doc_count': 0},\n",
       "   {'key': '239770.23', 'doc_count': 0},\n",
       "   {'key': '244663.50000000003', 'doc_count': 0},\n",
       "   {'key': '249556.77000000002', 'doc_count': 0},\n",
       "   {'key': '254450.04000000004', 'doc_count': 0},\n",
       "   {'key': '259343.31000000003', 'doc_count': 0},\n",
       "   {'key': '264236.58', 'doc_count': 0},\n",
       "   {'key': '269129.85000000003', 'doc_count': 0},\n",
       "   {'key': '274023.12', 'doc_count': 0},\n",
       "   {'key': '278916.39', 'doc_count': 0},\n",
       "   {'key': '283809.66000000003', 'doc_count': 0},\n",
       "   {'key': '288702.93000000005', 'doc_count': 0},\n",
       "   {'key': '293596.2', 'doc_count': 0},\n",
       "   {'key': '298489.47000000003', 'doc_count': 0},\n",
       "   {'key': '303382.74000000005', 'doc_count': 0},\n",
       "   {'key': '308276.01', 'doc_count': 0},\n",
       "   {'key': '313169.28', 'doc_count': 0},\n",
       "   {'key': '318062.55000000005', 'doc_count': 0},\n",
       "   {'key': '322955.82', 'doc_count': 0},\n",
       "   {'key': '327849.09', 'doc_count': 0},\n",
       "   {'key': '332742.36000000004', 'doc_count': 0},\n",
       "   {'key': '337635.63', 'doc_count': 0},\n",
       "   {'key': '342528.9', 'doc_count': 0},\n",
       "   {'key': '347422.17000000004', 'doc_count': 0},\n",
       "   {'key': '352315.44000000006', 'doc_count': 0},\n",
       "   {'key': '357208.71', 'doc_count': 0},\n",
       "   {'key': '362101.98000000004', 'doc_count': 0},\n",
       "   {'key': '366995.25000000006', 'doc_count': 0},\n",
       "   {'key': '371888.52', 'doc_count': 0},\n",
       "   {'key': '376781.79000000004', 'doc_count': 0},\n",
       "   {'key': '381675.06000000006', 'doc_count': 0},\n",
       "   {'key': '386568.33', 'doc_count': 0},\n",
       "   {'key': '391461.60000000003', 'doc_count': 0},\n",
       "   {'key': '396354.87000000005', 'doc_count': 0},\n",
       "   {'key': '401248.14', 'doc_count': 0},\n",
       "   {'key': '406141.41000000003', 'doc_count': 0},\n",
       "   {'key': '411034.68000000005', 'doc_count': 0},\n",
       "   {'key': '415927.95', 'doc_count': 0},\n",
       "   {'key': '420821.22000000003', 'doc_count': 0},\n",
       "   {'key': '425714.49000000005', 'doc_count': 0},\n",
       "   {'key': '430607.76', 'doc_count': 0},\n",
       "   {'key': '435501.03', 'doc_count': 0},\n",
       "   {'key': '440394.30000000005', 'doc_count': 0},\n",
       "   {'key': '445287.57000000007', 'doc_count': 0},\n",
       "   {'key': '450180.84', 'doc_count': 0},\n",
       "   {'key': '455074.11000000004', 'doc_count': 0},\n",
       "   {'key': '459967.38000000006', 'doc_count': 0},\n",
       "   {'key': '464860.65', 'doc_count': 0},\n",
       "   {'key': '469753.92000000004', 'doc_count': 0},\n",
       "   {'key': '474647.19000000006', 'doc_count': 0},\n",
       "   {'key': '479540.46', 'doc_count': 0},\n",
       "   {'key': '484433.73000000004', 'doc_count': 0},\n",
       "   {'key': '489327.00000000006', 'doc_count': 0},\n",
       "   {'key': '494220.27', 'doc_count': 0},\n",
       "   {'key': '499113.54000000004', 'doc_count': 0}],\n",
       "  'min': 10597.0,\n",
       "  'max': 10662.0},\n",
       " 'pos': {'frequency': [{'doc_count': 1, 'key': '10597'},\n",
       "   {'doc_count': 1, 'key': '10603'},\n",
       "   {'doc_count': 1, 'key': '10614'},\n",
       "   {'doc_count': 1, 'key': '10630'},\n",
       "   {'doc_count': 1, 'key': '10632'},\n",
       "   {'doc_count': 1, 'key': '10649'},\n",
       "   {'doc_count': 1, 'key': '10655'},\n",
       "   {'doc_count': 1, 'key': '10659'},\n",
       "   {'doc_count': 1, 'key': '10662'},\n",
       "   {'doc_count': 0, 'key': '10229'},\n",
       "   {'doc_count': 0, 'key': '10408'},\n",
       "   {'doc_count': 0, 'key': '10775'},\n",
       "   {'doc_count': 0, 'key': '11689'},\n",
       "   {'doc_count': 0, 'key': '11720'},\n",
       "   {'doc_count': 0, 'key': '11739'},\n",
       "   {'doc_count': 0, 'key': '11788'},\n",
       "   {'doc_count': 0, 'key': '11797'},\n",
       "   {'doc_count': 0, 'key': '11879'},\n",
       "   {'doc_count': 0, 'key': '11982'},\n",
       "   {'doc_count': 0, 'key': '11986'}],\n",
       "  'missing': {'doc_count': 0},\n",
       "  'doc_count': 9}}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "query myQuery {\n",
    "  GetAggsByChromosome(chr: \"2\", end: 100000, start: 10) {\n",
    "    chr {\n",
    "      histogram {\n",
    "        key\n",
    "        doc_count\n",
    "      }\n",
    "      min\n",
    "      max\n",
    "    }\n",
    "    pos {\n",
    "      frequency {\n",
    "        doc_count\n",
    "        key\n",
    "      }\n",
    "      missing {\n",
    "        doc_count\n",
    "      }\n",
    "      doc_count\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "response = requests.post(f\"{BASE_URL}{GRAPHQL_ENDPOINT}\", json={'query': query})\n",
    "\n",
    "data = json.loads(response.text)\n",
    "aggs_by_chromosome = data['data']['GetAggsByChromosome']\n",
    "aggs_by_chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a5f4610a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr.histogram</th>\n",
       "      <th>chr.min</th>\n",
       "      <th>chr.max</th>\n",
       "      <th>pos.frequency</th>\n",
       "      <th>pos.missing.doc_count</th>\n",
       "      <th>pos.doc_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[{'key': '9786.54', 'doc_count': 9}, {'key': '...</td>\n",
       "      <td>10597.0</td>\n",
       "      <td>10662.0</td>\n",
       "      <td>[{'doc_count': 1, 'key': '10597'}, {'doc_count...</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       chr.histogram  chr.min  chr.max  \\\n",
       "0  [{'key': '9786.54', 'doc_count': 9}, {'key': '...  10597.0  10662.0   \n",
       "\n",
       "                                       pos.frequency  pos.missing.doc_count  \\\n",
       "0  [{'doc_count': 1, 'key': '10597'}, {'doc_count...                      0   \n",
       "\n",
       "   pos.doc_count  \n",
       "0              9  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flattened_data = pd.json_normalize(aggs_by_chromosome, max_level=4)\n",
    "snp_df = pd.DataFrame(flattened_data)\n",
    "snp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ff7ac1a",
   "metadata": {},
   "source": [
    "# Download Snp Data by chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "8ae69605",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/downloads/22074ef2-6729-4839-8dc2-a51014f14382.txt'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "query myQuery {\n",
    "  DownloadSNPsByChromosome(\n",
    "    chr: \"2\"\n",
    "    end: 1000000\n",
    "    fields: [\"chr\", \"pos\", \"ref\"]\n",
    "    start: 10\n",
    "    page_args: {from_: 2, size: 2}\n",
    "  )\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "response = requests.post(f\"{BASE_URL}{GRAPHQL_ENDPOINT}\", json={'query': query})\n",
    "\n",
    "data = json.loads(response.text)\n",
    "download_by_chromosome = data['data']['DownloadSNPsByChromosome']\n",
    "download_by_chromosome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77dd36f8",
   "metadata": {},
   "source": [
    "# Count SNP by chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "602aaed0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "query myQuery {\n",
    "  CountSNPsByChromosome(\n",
    "    chr: \"2\"\n",
    "    end: 100000\n",
    "    start: 10\n",
    "    filter_args: {exists: [\"chr\", \"ref\", \"pos\"]}\n",
    "  )\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "response = requests.post(f\"{BASE_URL}{GRAPHQL_ENDPOINT}\", json={'query': query})\n",
    "\n",
    "data = json.loads(response.text)\n",
    "count_by_chromosome = data['data']['CountSNPsByChromosome']\n",
    "count_by_chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa19a2cd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
